ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/TransactionReaper.java

```diff
@@ -108,7 +108,18 @@ public final void check()
                 // TODO close window where first can change - maybe record nextDynamicCheckTime before probing first,
                 // then use compareAndSet? Although something will need to check before sleeping anyhow...
                 if (reaperElement == null) {
-                    nextDynamicCheckTime.set(Long.MAX_VALUE);
+                    // Reading _inShutdown in this synchronized block makes sure that its value does not come from
+                    // a cache (i.e. its value has been read from memory and not from the CPU cache)
+                    if (!_inShutdown)
+                        nextDynamicCheckTime.set(Long.MAX_VALUE);
+                    else {
+                        // If the JVM reaches the code then it means that the TransactionReaper is in shutdown and
+                        // either another thread has called removeElementClient but read a cached value for _inShutdown,
+                        // or the other thread is in the process of executing removeElementClient, and we preemptively
+                        // decide to notify here in case _inShutdown returns a stale value of false for that thread.
+                        // (ReaperThread will not wait at the next loop thus no need to use notifyAll())
+                        this.notify();
```

Comment:
I managed to get further in my debugger somehow but I am not hitting a breakpoint here still. I am thinking that by holding off the ReaperWorkerThread (so it doesn't do the notify as per the other discussion we are having) it might be preventing us reaching this point. What I am finding is there is a reaperElement still (albeit it seemed to be scheduled to be cancelled but it wasn't null). Basically, I am not seeing this line of code here that I think we want to see get executed, be executed.

Reply:
Hmm, I might be a bit wrong here. When I made it the only breakpoint I can see it triggered. I presume this could be being blocked from firing promptly by a sleep (https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/coordinator/ReaperThread.java#L63). But I think having seen it fire I will resolve this comment. But I do have what I think is the most significant point still which is that I think the terminate takes the length of the next timeout to stop. I will resolve this thread and I will continue that discussion in the test itself.

Reply:
> ... But I do have what I think is the most significant point still which is that I think the terminate takes the length of the next timeout to stop. I will resolve this thread and I will continue that discussion in the test itself.

Yep, that is correct.